File: 15.BackTracking/51.N-Queens.py
class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:
        board = [['.'] * n for _ in range (n)]
        result = []
        self.backtrack(board, 0, result)
        return result
    
    def backtrack(self, board, row, result):
        if row == len(board):
            # temp = []
            # for row in board:
            #     temp.append("".join(row))
            # result.append(temp)
            result.append(board.copy())
            return
        
        cols = len(board[row])
        for col in range(cols):
            if not self.is_valid(board, row, col):
                continue
            board[row][col] = "Q"
            self.backtrack(board, row + 1, result)
            board[row][col] = "."
    
    def is_valid(self, board, row, col):
        for i in range(row):
            if board[i][col] == "Q":
                return False
            
        for i, j in zip(range(row - 1, -1, -1), range(col - 1, -1, -1)):
            if board[i][j] == "Q":
                return False
            
        for i, j in zip(range(row - 1, -1, -1), range(col + 1, len(board))):
            if board[i][j] == "Q":
                return False
        return True

File: 15.BackTracking/51.N-Queens.cpp
#include<iostream>
#include<vector>
#include<string>

using namespace std;

class Solution{
    public:
        vector<vector<string>> solveNQueens(int n){
            vector<vector<string>> res;
            vector<string> nQueens(n, string(n, '.'));
            dfs(nQueens, res, 0, n);
            return res;
        }

        void dfs(vector<string>& nQueens, vector<vector<string>>& res, int row, int& n){
            if(row == n){                             // N层树
                res.push_back(nQueens);
                return;
            }
            for(int col = 0; col < n; ++col){         // N叉树
                if(isValid(nQueens, row, col, n)){    // 剪枝
                    nQueens[row][col] = 'Q';
                    dfs(nQueens, res, row + 1, n);
                    nQueens[row][col] = '.';
                }
            }
        }

        bool isValid(vector<string>& nQueens, int row, int col, int& n){
            // 检查坐标(row, col)在column（上半部）上是否存在'Q'
            for(int i = 0; i < row; ++i){
                if(nQueens[i][col] == 'Q') return false;
            }
            // 检查坐标(row, col)在45°对角线（上半部）上是否存在'Q'
            for(int i = row, j = col; i >= 0 && j < n; --i, ++j){
                if(nQueens[i][j] == 'Q') return false;
            }
            // 检查坐标(row, col)在135°对角线（上半部）上是否存在'Q'
            for(int i = row, j = col; i >= 0 && j >= 0; --i, --j){
                if(nQueens[i][j] == 'Q') return false;
            }
            return true;
        }
};

